hibernate-core/src/main/java/org/hibernate/query/sqm/tree/select/SqmSelectStatement.java

```diff
@@ -113,6 +118,27 @@ private SqmSelectStatement(
 		this.parameters = parameters;
 	}
 
+	/**
+	 * A query that returns the number of results of this query.
+	 *
+	 * @since 6.5
+	 */
+	@Incubating
+	public SqmSelectStatement<Long> countQuery() {
```

Comment:
Did you not see the existing `createCountQuery()` method a few lines below?

Reply:
So I looked at it. I'm definitely not sure, but I don't think it does quite what I want:

- I might be wrong, but the way it handles query parameters is, I guess, not quite right for this particular need. I believe I need to reuse the parameter instances of the original query, since the copy is to be executed via the same instance of `SelectionQuery`. (But I forget the subtleties here and I might be wrong!)
- It looks like it always uses the wrapping-in-a-subquery approach, which is safest and less bugprone, of course, but not very elegant&mdash;nor as efficient as I would prefer&mdash;for the case of simple queries.

So we should probably look into unifying these two implementations.

Reply:
>since the copy is to be executed via the same instance of SelectionQuery. (But I forget the subtleties here and I might be wrong!)

My code also does not copy parameters so it should just work fine.

>It looks like it always uses the wrapping-in-a-subquery approach, which is safest and less bugprone, of course, but not very elegant—nor as efficient as I would prefer—for the case of simple queries.

I don't think there is any real database that executes the two queries i.e. wrapped in `select count(*) from (..)` and your in-line variant differently, so that point is a bit moot.
I don't mind if you alter the `createCountQuery()` method to behave this way though i.e. your in-line count variant.

Reply:
> I don't think there is any real database that executes the two queries i.e. wrapped in select count(*) from (..) and your in-line variant differently, so that point is a bit moot.

I mean, it's definitely different, since I'm trimming out unnecessary left joins.

In your code, how do you handle getting the correct result count (rather than _row count_) for a query like `from Book left join fetch authors`?

Reply:
Oh well, result count is a different topic, I didn't understand at first that is what you want to achieve. The query generated by `creatCountQuery` produces the row count, though I don't know if it should. I guess the result count is more useful, so maybe you should do these changes in `creatCountQuery`.

Reply:
(This is why I thought you were copying the parameters, by the way.)

Reply:
Queries that were constructed from `HQL` have the `parameters` set to something non-null, but Criteria queries have `null` set, because the parameters always need to be recomputed.

Reply:
OK so I need that code when the criteria API is the source of the SQM tree?